src/ray/raylet/worker_pool.cc

```diff
@@ -968,6 +970,10 @@ void WorkerPool::PushWorker(const std::shared_ptr<WorkerInterface> &worker) {
       // TODO(SongGuyang): This worker will not be used forever. We should kill it.
       state.idle_dedicated_workers[task_id] = worker;
     }
+    // We either have an idle worker or a slot to start a new worker.
+    if (worker->GetWorkerType() == rpc::WorkerType::WORKER) {
+      TryPendingPopWorkerRequests(worker->GetLanguage());
```

Comment:
Better to invoke this function in [OnWorkerStarted](https://github.com/ray-project/ray/blob/cff74697fd1cfa50b2c1eb098f85bd7aba0831bb/src/ray/raylet/worker_pool.cc#L739) ? I suggest this because the flag `is_pending_registration` is set to `false` in this function and it is related to the concurrent starting worker process count.

Reply:
We don't only retry pending pop worker requests when `is_pending_registration` is set to `false` but also when an idle worker is returned. `PushWorker()` handles both case.

If we do it in `OnWorkerStarted()` then we may start more workers than necessary. Imagining a pre-started worker is registered, `OnWorkerStarted()` is called first, if we retry in this function, then we will start a new worker (since the new worker is not pushed yet) but if we retry in `PushWorker` then we can simply use the pre-started worker without starting a new one.